examples/advanced/nce_example.py

```diff
@@ -0,0 +1,237 @@
+#!/usr/bin/python3
+"""
+Counterfactual explanation of a user query.
+
+Based on application on the knapsack problem of Korikov, A., & Beck, J. C. Counterfactual Explanations via Inverse Constraint Programming.
+
+Usecase:
+ 1) Some optimal solution x* is provided to the user by a constraint optimization solver.
+ 2) User enters a query to change solution variables x_i to a new value.
+    This extra set of constraints is used to generate a new optimal solution x_d
+ 3) Using inverse optimization, we calculate the values of constraints the objective function which result in x_d being the optimal solution.
+    Extra constraint: only values corresponding to foil items are allowed to change.
+ 4) The user is presented with the result.
+
+
+Given:
+    A constraint optimization problem (c, f, X) with an optimal solution x*
+    A set of foil constraints assigning a value to variables in x*, resulting in x_d
+Find:
+    d such that min(f(d,X)) == x_d and ||c - d|| is minimal
+
+
+The algorithm consists of 2 alternating problems: the master problem (MP), and the sub problem (SP)
+
+Algorithm:
+    1 S = {}
+    2 Solve MP to obtain optimal d*
+    3 Solve SP with d* as c to obtain x_0
+    4 if objective(x_d) > objective(x_0):
+        add x_0 to S
+        go to 2
+      else:
+          return d*
+
+
+Master problem:
+    Constraints:
+        - constraints of original forward problem
+        - objective(d*, x) >= objective(d*, x_0) forall x_0 in S
+        - foil constraints assigning values to x_i
+        - foil contraints restricting the altering of c to match the foil constraint indices
+    Objective:
+        - Minimize || c - d* ||_1
+
+Sub problem:
+    The original forward problem
+
+
+"""
+
+from cpmpy import *
+import numpy as np
+
+INFINITY = np.iinfo(np.int32).max
+verbose = False
+np.random.seed(0)
+
+def main():
+
+    n = 10  #Number of items in the knapsack
+    m = 5   #Number of items to change in the knapsack
+
+    values, weights, capacity, x = generate_knapsack_model(n)
+    #TODO: verbose statement?
+    print_knapsack_model(values, weights, capacity, x)
+
+    x_user, foil_items = generate_foil_knapsack(values, weights, capacity, x, m)
```

Comment:
a comment like: # we emulate a foil query by a hypothetical user, our generation process ensures that...